### PR TITLE
Update: Replace polldaddy shortcode with crowdsignal

### PR DESCRIPTION
**Pull request by ice9js**
As a first step for pabtAt-r-p2, this PR introduces a new `crowdsignal` shortcode which takes over the functionality of `polldaddy`. The latter will remain functional as an alias to `crowdsignal`.  

#### Testing instructions:

You should be able to embed polls/surveys and ratings through both `polldaddy` and `crowdsignal` shortcodes.

#### Proposed changelog entry for your changes:

Replaces `polldaddy` shortcode with `crowdsignal`.

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
What's your timeline for this? Would you like this to be included in the next release, scheduled to be released at the beginning of next month?

Other than that, I would recommend that we leave the `modules/shortcodes/polldaddy.php` as a stub, to avoid issues in some environments. See https://github.com/Automattic/jetpack/blob/master/modules/debug.php

Other than that, the changes seem to work well. 👍

**Comment by ice9js** [comment]
>What's your timeline for this? Would you like this to be included in the next release, scheduled to be released at the beginning of next month?

Our current approach is to get everything in place before deciding on an exact launch date.  
I think I'd be most comfortable with making a few more PR's off this branch - for easier review as all other work is based on it - and merge everything before the launch.

We don't want to update any public references before the launch and I think it'll be easier to manage everything together.

>Other than that, I would recommend that we leave the modules/shortcodes/polldaddy.php as a stub, to avoid issues in some environments.

Done 😄

**Review by dbtlr — DISMISSED** [review]

**Comment by jeherve** [comment]
Is this ready for a review now?

**Comment by ice9js** [comment]
Hey @jeherve! Yeah, it's ready.

**Comment by jeherve** [comment]
Thanks! It looks like this will need a rebase before it can be tested and merged though.

Thank you!

**Comment by ice9js** [comment]
I rebased the branch - should be all right now :)

**Review by jeherve — CHANGES_REQUESTED** [review]
This looks good, I only have a few remarks.

File: modules/shortcodes/crowdsignal.php [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+		return jetpack_preg_replace_outside_tags( '!(?:\n|\A)http://(polldaddy.com/poll|poll.fm)/([0-9]+?)/(.+)?(?:\n|\Z)!i', "\n<script type='text/javascript' charset='utf-8' async src='//static.polldaddy.com/p/$1.js'></script><noscript> <a href='http://polldaddy.com/poll/$1/'>View Poll</a></noscript>\n", $content, 'polldaddy.com/poll' );
+	}
+
+	// higher priority because we need it before auto-link and autop get to it
+	add_filter( 'the_content', 'crowdsignal_link', 1 );
+	add_filter( 'the_content_rss', 'crowdsignal_link', 1 );
+}
+
+wp_oembed_add_provider( '#https?://.*\.survey\.fm/.*#i', 'https://api.crowdsignal.com/oembed', true );
+wp_oembed_add_provider( '#https?://poll\.fm/.*#i', 'https://api.crowdsignal.com/oembed', true );
```
[the remark this answers is not in the document]
Comment (ice9js) [reply modules/shortcodes/crowdsignal.php]: Yeah, this is done as part of pMz3w-96r-p2.

File: modules/shortcodes/crowdsignal.php [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+	function crowdsignal_link( $content ) {
+		return jetpack_preg_replace_outside_tags( '!(?:\n|\A)http://(polldaddy.com/poll|poll.fm)/([0-9]+?)/(.+)?(?:\n|\Z)!i', "\n<script type='text/javascript' charset='utf-8' async src='//static.polldaddy.com/p/$1.js'></script><noscript> <a href='http://polldaddy.com/poll/$1/'>View Poll</a></noscript>\n", $content, 'polldaddy.com/poll' );
+	}
+
+	// higher priority because we need it before auto-link and autop get to it
+	add_filter( 'the_content', 'crowdsignal_link', 1 );
+	add_filter( 'the_content_rss', 'crowdsignal_link', 1 );
+}
+
+wp_oembed_add_provider( '#https?://.*\.survey\.fm/.*#i', 'https://api.crowdsignal.com/oembed', true );
```
[the remark this answers is not in the document]
Comment (ice9js) [reply modules/shortcodes/crowdsignal.php]: Thank you for asking this! This is exactly the reason why I added the other two as well.
I initially skipped `polldaddy.com` as those links will become deprecated but considering this again - we still need this compatibility.

I added the missing provider.

File: modules/shortcodes/crowdsignal.php [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+	}
+}
+
+// kick it all off
+new CrowdsignalShortcode();
+
+if ( ! function_exists( 'crowdsignal_link' ) ) {
+	// http://polldaddy.com/poll/1562975/?view=results&msg=voted
+	function crowdsignal_link( $content ) {
+		return jetpack_preg_replace_outside_tags( '!(?:\n|\A)http://(polldaddy.com/poll|poll.fm)/([0-9]+?)/(.+)?(?:\n|\Z)!i', "\n<script type='text/javascript' charset='utf-8' async src='//static.polldaddy.com/p/$1.js'></script><noscript> <a href='http://polldaddy.com/poll/$1/'>View Poll</a></noscript>\n", $content, 'polldaddy.com/poll' );
```
[the remark this answers is not in the document]
Comment (ice9js) [reply modules/shortcodes/crowdsignal.php]: Thanks for noticing this! Embed JavaScript will still be served from `static.polldaddy.com` but we should redirect to `poll.fm`.  
I also updated the expression to correctly match & replace both `polldaddy.com` and `poll.fm` links. For now I used `preg_replace` for better readability but can switch back to nested `jetpack_preg_replace_outside_tags` calls if you think performance impact will be too much.

**Review by jeherve — DISMISSED** [review]
It seems the changes broke some of our shortcodes, when using a custom domain. Here is an example:

`[polldaddy type="iframe" survey="EB151947E5950FCF" height="auto" domain="jeherve" id="a-survey-with-branches"]`

screenshot 2018-10-15 at 20 56 20

**Comment by ice9js** [comment]
Thanks for spotting that! Should be taken care of now :)

**Review by jeherve — CHANGES_REQUESTED** [review]
I seem to have trouble displaying polls when they are added with just a link and rely on oEmbed. Here is the JS error I get:

screenshot 2018-10-17 at 12 01 38

I only inserted this in the editor:

`https://polldaddy.com/poll/7910844/`

**Comment by jeherve** [comment]
The other shortcodes and embed methods appear to work well; for some reason only this seems to fail. I am afraid I don't have much more info about it. It is on an installation of WordPress that uses the latest Stable version of Core, and Jetpack with your branch.

screenshot 2018-10-17 at 13 53 21

It seems to fail because of the error above:

> Failed to execute 'write' on 'Document': It isn't possible to write into a document from an asynchronously-loaded external script unless it is explicitly opened.

I have nothing else in the post content. Here is the output:

```html

<div class="entry-content">
<p><script type='text/javascript' charset='utf-8' async src='//static.polldaddy.com/p/7910844.js'></script><noscript> <a href='https://poll.fm/7910844'>View Poll</a></noscript></p>
</div><!-- .entry-content -->
```

This happens in Chrome.

**Comment by ice9js** [comment]
For some reason I thought the issue was within the editor itself, not the published post. Indeed removing `async` from the `<script>` tag seems to have helped and that's what I went with for now.

Eventually this should be addressed in Crowdsignal itself, however due to pabtAt-1f-p2 that will probably take a while.

**Review by jeherve — APPROVED** [review]
This looks good to me now, I've tried all the formats I could think of and did not run into issues. I would love if someone else from your team could test and review as well, though!

Thank you!

**Comment by kraftbj** [comment]
Once we're finally ready, we'll need to take over and manually apply the patches to D18089-code, including updating the build script to the new file paths while spreading it over a couple of commits.

**Comment by kraftbj** [comment]
@ice9js - If everything is ready, could you have someone more familiar with the changeover give this a look? I'm manually marking the WP.com sync test as successful and I can work on doing the actual sync work once this lands.

**Comment by ice9js** [comment]
@kraftbj - apologies for the delay there. We currently have our hands full with pabtAt-2q-p2.  
Once that's taken care of @dbtlr or @donnchawp will have a look later this week. We definitely still do want to get this out with the next release.

**Comment by kraftbj** [comment]
@ice9js Understood. Tuesday the 30th is code freeze for anything not regressions, so trying to give y'all time to make sure this makes it in.

**Comment by ice9js** [comment]
Thanks for the heads up!

**Comment by dbtlr** [comment]
We're ready to go live with the next Jetpack release. What can we do to get this merged with wpcom?

**Comment by jeherve** [comment]
@dbtlr Could someone on your team review and test this PR, so we can get this merged and included in the next release?

Thank you!

**Review by dbtlr — APPROVED** [review]
LGTM

**Comment by kraftbj** [comment]
Noting that @donnchawp is handling the changes on the WP.com side to match this. I'll coordinate with him to make sure Fusion is tracking things once the dust settles.

**Pull request merged**